// @voximplant/voxengine-ci/lib/domains/entities/vox-scenario.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoxScenarioBuilderTsConfig = exports.VoxScenarioMetadata = exports.VoxScenario = void 0;
class VoxScenario {
    constructor(scenarioInfo) {
        this.scenarioName = scenarioInfo.scenarioName;
        this.scenarioScript = scenarioInfo.scenarioScript ?? '';
    }
}
exports.VoxScenario = VoxScenario;
class VoxScenarioMetadata {
    constructor(scenarioInfo, hash) {
        this.scenarioId = scenarioInfo.scenarioId;
        this.scenarioName = scenarioInfo.scenarioName;
        this.hash = hash;
    }
}
exports.VoxScenarioMetadata = VoxScenarioMetadata;
class VoxScenarioBuilderTsConfig {
    constructor(outDir, include, exclude) {
        this.compilerOptions = {
            allowJs: true,
            target: 'ES2020',
            noEmitOnError: true,
            skipLibCheck: true,
            noLib: true,
            outDir: outDir,
        };
        this.include = include;
        this.exclude = exclude;
    }
}
exports.VoxScenarioBuilderTsConfig = VoxScenarioBuilderTsConfig;
